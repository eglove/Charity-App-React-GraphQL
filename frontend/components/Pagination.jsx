import React from "react";
import gql from "graphql-tag";
import {Query} from "react-apollo";
import Head from 'next/head';
import Link from 'next/link';
import PaginationStyles from "./styles/PaginationStyles";
import {perPage} from "../config";

const PAGINATION_QUERY = gql`
    query PAGINATION_QUERY {
        charitiesConnection {
            aggregate {
                count
            }
        }
    }
`;

const Pagination = props => (
    <Query query={PAGINATION_QUERY}>
        {({data, loading, error}) => {
            if (loading) return <p>Loading...</p>;
            const count = data.charitiesConnection.aggregate.count;
            const pages = Math.ceil(count / perPage);
            const currentPage = props.page;
            return (
                <>
                    <Head>
                        <title>Cognitame ðŸ“„ Page {currentPage} of {pages}</title>
                    </Head>
                    <PaginationStyles>
                        <Link prefetch href={{
                            pathname: '/charities',
                            query: {page: currentPage - 1}
                        }}>
                            <a className="prev" aria-disabled={currentPage <= 1}>â—€</a>
                        </Link>
                        <p>Page {currentPage} of {pages} Pages</p>
                        <p>ðŸ’™ {count} Charities Total ðŸ’™</p>
                        <Link prefetch href={{
                            pathname: '/charities',
                            query: {page: currentPage + 1}
                        }}>
                            <a className="next" aria-disabled={currentPage >= pages}>â–¶</a>
                        </Link>
                    </PaginationStyles>
                </>
            );
        }}
    </Query>
);

export default Pagination;