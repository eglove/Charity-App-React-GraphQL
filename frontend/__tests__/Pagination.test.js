import {mount} from "enzyme";
import wait from "waait";
import toJson from "enzyme-to-json";
import Router from "next/router";
import Pagination, {PAGINATION_QUERY} from "../components/Pagination";
import {MockedProvider} from "react-apollo/test-utils";

Router.router = {
    push() {
    },
    prefetch(url) {
    },
};

function makeMocksFor(length) {
    return [
        {
            request: {query: PAGINATION_QUERY},
            result: {
                data: {
                    charitiesConnection: {
                        __typename: 'aggregate',
                        aggregate: {
                            __typename: 'count',
                            count: length
                        },
                    },
                },
            },
        },
    ];
}

describe('<Pagination/>', () => {
    it('displays a loading message', () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(1)}>
                <Pagination page={1}/>
            </MockedProvider>
        );
        expect(wrapper.text()).toContain('Loading...');
    });

    it('renders pagination for 18 items', async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={1}/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        expect(wrapper.find('.totalPages').text()).toEqual('5');
        const nav = wrapper.find('PaginationStyles');
        expect(toJson(nav)).toMatchSnapshot();
    });

    it('disables prev button on first page', async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(12)}>
                <Pagination page={1}/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(true);
        expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
    });

    it('disables next button on last page', async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(12)}>
                <Pagination page={3}/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(true);
        expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
    });

    it('enables all buttons on middle page', async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(12)}>
                <Pagination page={2}/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
        expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
    });
});
