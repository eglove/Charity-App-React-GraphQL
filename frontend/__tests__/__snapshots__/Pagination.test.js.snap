// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> renders pagination for 18 items 1`] = `
<PaginationStyles>
  <div
    className="PaginationStyles-sz1p4z-0 gGNYCi"
  >
    <Link
      href={
        Object {
          "pathname": "/charities",
          "query": Object {
            "page": 0,
          },
        }
      }
      prefetch={true}
    >
      <a
        aria-disabled={true}
        className="prev"
        href="/charities?page=0"
        onClick={[Function]}
      >
        â—€
      </a>
    </Link>
    <p>
      Page 
      1
       of 
      <span
        className="totalPages"
      >
        5
      </span>
       Pages
    </p>
    <p>
      ðŸ’™ 
      18
       Charities Total ðŸ’™
    </p>
    <Link
      href={
        Object {
          "pathname": "/charities",
          "query": Object {
            "page": 2,
          },
        }
      }
      prefetch={true}
    >
      <a
        aria-disabled={false}
        className="next"
        href="/charities?page=2"
        onClick={[Function]}
      >
        â–¶
      </a>
    </Link>
  </div>
</PaginationStyles>
`;
